worlds/AutoWorld.py

```diff
@@ -365,133 +364,311 @@ def stage_assert_generate(cls, multiworld: "MultiWorld") -> None:
-        pass
+    @classmethod
+    def stage_pre_fill(cls, multiworld: "MultiWorld") -> None:
+        """
+        Class level stage of pre_fill. Gets run once per world type per multiworld after all the instanced calls.
+
+        :param multiworld: The multiworld object for this generation.
+        """
+
+    def fill_hook(
+            self,
+            progitempool: list["Item"],
+            usefulitempool: list["Item"],
+            filleritempool: list["Item"],
+            fill_locations: list["Location"],
+    ) -> None:
+        """
+        Special method that gets called as part of distribute_items_restrictive (main fill).
+
+        :param prog_item_pool: The pool of progression items to be placed.
+        :param useful_item_pool: The pool of useful items to be placed.
+        :param filler_item_pool: The pool of filler items to be placed.
+        :param fill_locations: All locations that still need to be filled.
+        """
+
+    @classmethod
+    def stage_fill_hook(
+            cls,
+            multiworld: "MultiWorld",
+            progitempool: list["Item"],
+            usefulitempool: list["Item"],
+            filleritempool: list["Item"],
+            fill_locations: list["Location"],
+    ) -> None:
+        """
+        Class level stage of fill_hook. Gets run once per world type per multiworld after all the instanced calls.
+
+        :param multiworld: The multiworld object for this generation.
+        :param prog_item_pool: The pool of progression items to be placed.
+        :param useful_item_pool: The pool of useful items to be placed.
+        :param filler_item_pool: The pool of filler items to be placed.
+        :param fill_locations: All locations that still need to be filled.
+        """
 
     def post_fill(self) -> None:
         """
         Optional Method that is called after regular fill. Can be used to do adjustments before output generation.
         This happens before progression balancing, so the items may not be in their final locations yet.
         """
 
+    @classmethod
+    def stage_post_fill(cls, multiworld: "MultiWorld") -> None:
+        """
+        Class level stage of post_fill. Gets run once per world type per multiworld after all the instanced calls.
+
+        :param multiworld: The multiworld object for this generation.
+        """
+
+    @classmethod
+    def stage_generate_output(cls, multiworld: "MultiWorld", output_directory: str) -> None:
+        """
+        Class level stage of generate_output. Gets run once per world type per multiworld before the instanced calls. As
+        this method gets called from a threadpool, do not use multiworld.random here. If any last-second randomization
+        needs to be done, use one of your worlds' random or seed a new random object. As these calls are all threaded,
+        there is no guarantee this method will complete before the instance methods unless the other threads await it.
```

Comment:
If I'm reading this right, this and some of the spoiler log methods are the only world methods where stage_x() is called _before_ x()? If so, it might be worth emphasizing that more, or mentioning _why_ these are exceptions.